### ISSUE TITLE
StreamProducer defer patch

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2736]
|<img alt="KyleJ's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| KyleJ reported|
|-|-|
|Trac ID|trac#2736|
|Type|defect|
|Created|2007-07-18 04:15:44Z|

If a callback occurs after self.deferred is set to None in _doWrite (for instance: while the producer is paused) and returns a deferred None, StreamProducder halts early.

Patch (for trunk/twisted/web2/stream.py):
```
--- stream.py   2007-07-17 21:08:26.000000000 -0700
+++ stream2.py  2007-07-17 21:09:12.000000000 -0700
@@ -729,7 +729,7 @@
     
     def resumeProducing(self):
         self.paused = False
-        if self.deferred is not None:
+        if self.deferred is not None and self.deferred.result is not None:
             return
 
         try:
```

<details><summary>Searchable metadata</summary>

```
trac-id__2736 2736
type__defect defect
reporter__KyleJ KyleJ
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__web2 web2
keywords__patch patch
time__1184732144000000 1184732144000000
changetime__1297621667000000 1297621667000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Got a unit test?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set status to `closed`|
|-|-|

Not going to fix this in web2 (see [#4821](https://github.com/twisted/twisted/issues/4821)).  Also I still don't quite understand how to trigger this behavior either.  There are no streams in Twisted Web, so no parallel misbehavior, so not retargeting this at the web component.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
